BYOC: Support template build for team clusters

sitole opened the pull request:
Implements abstraction for template managers in API to support edge clusters.

sitole commented on packages/api/internal/template-manager/template_manager.go:
```diff
@@ -71,247 +82,105 @@ func (tm *TemplateManager) BuildsStatusPeriodicalSync(ctx context.Context) {
+			// todo: we need to get team here soe i can get cludster id
 			zap.L().Info("Running periodical sync of builds statuses", zap.Int("count", len(buildsRunning)))
-			for _, buildDB := range buildsRunning {
-				go tm.BuildStatusSync(ctx, buildDB.ID, *buildDB.EnvID)
-			}
+			//for _, buildDB := range buildsRunning {
+			//
+			//	buildDB.
+			//
+			//
+			//	go tm.BuildStatusSync(ctx, buildDB.ID, *buildDB.EnvID, nil, buildDB.ClusterNodeID)
+			//}
```
This needs to be adjusted

sitole commented on packages/api/internal/grpc/client.go:
```diff
@@ -0,0 +1,28 @@
+	tempaltemanagergrpc "github.com/e2b-dev/infra/packages/shared/pkg/grpc/template-manager"
+)
+
+type GRPCClient struct {
+	Info     infogrpc.InfoServiceClient
+	Sandbox  orchestratorgrpc.SandboxServiceClient
+	Template tempaltemanagergrpc.TemplateServiceClient
+
+	Connection *grpc.ClientConn
+}
+
+func (a *GRPCClient) Close() error {
+	err := a.Connection.Close()
+	if err != nil {
+		return fmt.Errorf("failed to close connection: %w", err)
+	}
+
+	return nil
+}
```
Shared GRPC client for template manager and orchestrator so we can use them in unified way across codebase

sitole commented on packages/shared/pkg/schema/build.go:
```diff
@@ -43,6 +43,7 @@ func (EnvBuild) Fields() []ent.Field {
 		field.String("kernel_version").Default(DefaultKernelVersion).SchemaType(map[string]string{dialect.Postgres: "text"}),
 		field.String("firecracker_version").Default(DefaultFirecrackerVersion).SchemaType(map[string]string{dialect.Postgres: "text"}),
 		field.String("envd_version").SchemaType(map[string]string{dialect.Postgres: "text"}).Nillable().Optional(),
+		field.String("cluster_node_id").SchemaType(map[string]string{dialect.Postgres: "text"}).Optional().Nillable(),
```
This is just to identify the node inside the cluster where the build runs. When we are taking a node, we are already scoped just to a specific cluster.

dobrac reviewed (changes requested):
I would mainly unite the syncing logic, as it's there now twice, but doing basically the same thing

sitole commented on packages/api/internal/edge/cluster.go:
```diff
@@ -0,0 +1,193 @@
+	}
+
+	if endpointTls {
+		cred := credentials.NewTLS(&tls.Config{MinVersion: tls.VersionTLS12})
```
By default, AWS load balancer with TLS is using version 12. I can manually adjust it.

sitole commented on packages/api/internal/edge/cluster_nodes.go:
```diff
@@ -0,0 +1,168 @@
+	// fetch cluster nodes with use of service discovery
+	res, err := c.httpClient.V1ServiceDiscoveryGetOrchestratorsWithResponse(c.ctx)
+	if err != nil {
+		zap.L().Error("Failed to get cluster nodes", zap.Error(err), l.WithClusterID(c.Id))
```
Where's it logged? Propably we should change it to something like `Failed to get cluster nodes from service discovery`

dobrac replied:
on line 53 in this file

sitole commented on packages/api/internal/template-manager/template_status.go:
```diff
@@ -0,0 +1,249 @@
+
+	envBuildDb, err := tm.db.GetEnvBuild(childCtx, buildID)
+	if err != nil {
+		logger.Error("Error when fetching env build for background sync", zap.Error(err))
```
This method is expected to run in a goroutine, so during initial we can add an anonymous go function and catch the returned error, but not sure if it will be cleaner. In this exact case, where function is called in the background.

dobrac replied:
I think if you try to test this method would clarify it a bit

sitole commented on packages/api/internal/edge/cluster.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+func (c *Cluster) GetAvailableTemplateBuilder() (*ClusterNode, error) {
+	_, span := c.tracer.Start(c.ctx, "template-builder-get-available-node")
```
You want to ideally know how long it takes to get an available builder per cluster.

dobrac replied:
but this will track all spans to the parent context/span, not span per `GetAvailableTemplateBuilder` invocation

sitole commented on packages/api/internal/edge/cluster_nodes.go:
```diff
@@ -0,0 +1,168 @@
+
+		// synchronize node state
+		if found {
+			wg.Add(1)
+			go func() {
+				c.syncNode(ctx, node)
+				wg.Done()
+			}()
```
true, removing

sitole commented on packages/api/internal/edge/cluster.go:
```diff
@@ -0,0 +1,193 @@
+)
+
+func NewCluster(ctx context.Context, tracer trace.Tracer, tel *telemetry.Client, endpoint string, endpointTls bool, secret string, id uuid.UUID) (*Cluster, error) {
+	// so we during cluster disconnect we don't cancel the upper context
```
Maybe the comment is just confusing. We are creating a ctx with a cancel for each sandbox. Then, during cluster close, ctx cancel is called, so all in-progress requests and synchronizations are shut down.

dobrac replied:
I understand the reason, it's just the comment sentence doesn't make much sense to me, nor explain fully the reason behind it

sitole replied:
okay, will remove it

dobrac commented on packages/api/internal/template-manager/template_manager.go:
```diff
@@ -39,22 +66,63 @@ const (
+)
+
+func New(ctx context.Context, tracer trace.Tracer, tracerProvider trace.TracerProvider, meterProvider metric.MeterProvider, db *db.DB, sqlcDB *sqlcdb.Client, edgePool *edge.Pool, lokiClient *loki.DefaultClient, buildCache *templatecache.TemplatesBuildCache) (*TemplateManager, error) {
+	conn, err := grpc.NewClient(templateManagerHost,
```
as the init of grpc is not just a single line/function call, I would split it to at least some helper function. Another approach might be to do it all in one/two line/function calls

sitole replied:
done, make sense

sitole commented on packages/api/internal/edge/cluster_nodes.go:
```diff
@@ -0,0 +1,168 @@
+			zap.L().Info("Cluster nodes sync ended", l.WithClusterID(c.Id))
+			return
+		case <-timer.C:
+			syncTimeout, syncCancel := context.WithTimeout(c.ctx, clusterNodesSyncInterval)
```
`c.ctx` removed and replaced with channel

sitole commented on packages/api/internal/edge/pool.go:
```diff
@@ -0,0 +1,163 @@
+	wg.Wait()
+
+	return nil
+}
```
introduces new utility for synchronization, solving this

sitole commented:
- Make consisten variable naming `templateId` -> `templateID` in all newly written logic and logic that we are touching here
- API is now using `NodeID ` that represents VM machine identification instead of `ServiceInstanceID` that represents edge/orchestrator process UUID. This is mainly because API is alrady using `NodeID` for orchestrators pool so we want to stay consistent
- Renamed edge API interface to better communicate differences between node id and service instance ids.
- Changed implementation of synchronization util that does not require map

sitole commented on packages/api/internal/edge/cluster_nodes.go:
```diff
@@ -113,46 +105,56 @@ func (d clusterSynchronizationStore) SourceList(ctx context.Context) ([]api.Clus
 
-func (d clusterSynchronizationStore) SourceKey(item api.ClusterOrchestratorNode) string {
-	return item.Id
+func (d clusterSynchronizationStore) SourceExists(ctx context.Context, s []api.ClusterOrchestratorNode, p *ClusterNode) bool {
```
nope for reason, fixing. thx!

sitole replied:
no need for that, structure should be self sustaining and different invocations should not do changes on shared state

sitole commented on packages/shared/pkg/synchronization/synchronization.go:
```diff
@@ -75,71 +74,74 @@ func (s *Synchronize[SourceItem, SourceKey, PoolItem]) SyncInBackground(cancel c
-func (s *Synchronize[SourceItem, SourceKey, PoolItem]) runSyncDiscovered(ctx context.Context, sourceItems []SourceItem) {
-	spanCtx, span := s.Tracer.Start(ctx, s.getSpanName("sync-discovered-items"))
+func (s *Synchronize[SourceItem, PoolItem]) Close() {
+	close(s.cancel)
```
will add `sync.Once`

sitole commented on packages/api/internal/edge/pool.go:
```diff
@@ -73,6 +72,8 @@ func (p *Pool) GetClusterById(id uuid.UUID) (*Cluster, bool) {
+
 	// Close pool, this needs to be called before closing the clusters
 	// so background jobs syncing cluster will not try to update the pool nodes
 	close(p.close)
```
removing, thx
